lighthouse-core/gather/computed/main-thread-tasks.js

```diff
@@ -0,0 +1,215 @@
+      group: taskGroups.other,
+      attributableURL: undefined,
+      duration: NaN,
+      selfTime: NaN,
+    };
+
+    if (parent) {
+      parent.children.push(newTask);
+    }
+
+    return newTask;
+  }
+
+  /**
+   * @param {LH.TraceEvent[]} traceEvents
+   * @return {TaskNode[]}
+   */
+  static _createTasksFromEvents(traceEvents) {
+    const {startedInPageEvt} = TraceProcessor.findTracingStartedEvt(traceEvents);
+
+    /** @type {TaskNode[]} */
+    const tasks = [];
+    /** @type {TaskNode|undefined} */
+    let currentTask;
+
+    for (const event of traceEvents) {
+      // Only look at main thread events
+      if (event.pid !== startedInPageEvt.pid || event.tid !== startedInPageEvt.tid) continue;
+      // Only look at X (Complete), B (Begin), and E (End) events as they have most data
+      if (event.ph !== 'X' && event.ph !== 'B' && event.ph !== 'E') continue;
+
+      // Update currentTask based on the elapsed time.
+      // The next event may be after currentTask has ended.
+      while (
+        currentTask &&
+        Number.isFinite(currentTask.endTime) &&
+        currentTask.endTime <= event.ts
+      ) {
+        currentTask = currentTask.parent;
+      }
+
+      // If we don't have a current task, start a new one.
+      if (!currentTask) {
+        // We can't start a task with an end event
+        if (event.ph === 'E') {
+          throw new Error('Fatal trace logic error');
+        }
+
+        currentTask = MainThreadTasks._createNewTaskNode(event);
+        tasks.push(currentTask);
+
+        continue;
+      }
+
+      if (event.ph === 'X' || event.ph === 'B') {
+        // We're starting a nested event, create it as a child and make it the currentTask
+        const newTask = MainThreadTasks._createNewTaskNode(event, currentTask);
+        tasks.push(newTask);
+        currentTask = newTask;
+      } else {
+        if (currentTask.event.ph !== 'B') {
+          throw new Error('Fatal trace logic error');
+        }
+
+        // We're ending an event, update the end time and the currentTask to its parent
+        currentTask.endTime = event.ts;
+        currentTask = currentTask.parent;
+      }
+    }
+
+    return tasks;
+  }
+
+  /**
+   * @param {TaskNode} task
+   * @return {number}
+   */
+  static _computeRecursiveSelfTime(task) {
+    const childTime = task.children
+      .map(MainThreadTasks._computeRecursiveSelfTime)
+      .reduce((sum, child) => sum + child, 0);
+    task.duration = task.endTime - task.startTime;
+    task.selfTime = task.duration - childTime;
+    return task.duration;
+  }
+
+  /**
+   * @param {TaskNode} task
+   * @param {string} [parentURL]
+   */
+  static _computeRecursiveAttributableURL(task, parentURL) {
+    const argsData = task.event.args.data || {};
+    const stackFrames = argsData.stackTrace || [{url: undefined}];
+    const taskURL = argsData.url || (stackFrames[0] && stackFrames[0].url);
+
+    task.attributableURL = parentURL || taskURL;
```

Comment:
why prefer a parent's URL? 

parseHTML will be an .html file but might have a scriptEvaluation child, which should be attributed to the script, right?

Reply:
my rationale was for FunctionCall within an EvaluateScript (i.e. /ad.js spends 1500ms calling lots of /jquery.js functions, and I want to know that /ad.js is the bad one, not /jquery.js), perhaps we just create a stack for attributableURLs and in bootup-time we find the first JS one/fallback to HTML if not exists?

Reply:
> , perhaps we just create a stack for attributableURLs and in bootup-time we find the first JS one/fallback to HTML if not exists?

sg.

yah lets